Modules/_threadmodule.c

```diff
@@ -1357,10 +1530,11 @@ thread_run(void *boot_raw)
     _PyThreadState_DeleteCurrent(tstate);
 
 exit:
-    if (thread_is_exiting != NULL) {
-        _PyEvent_Notify(&thread_is_exiting->event);
-        _PyEventRc_Decref(thread_is_exiting);
-    }
+    // Don't need to wait for this thread anymore
+    remove_from_shutdown_handles(handle);
```

Comment:
I think there's a possible race condition here between destruction of the `thread_module_state` during interpreter shutdown and this call. The `thread_module_state` containing the head of the linked list could already be deallocated. We can't fully rely on `thread_shutdown` because that can be interrupted by ctrl-c.

One possibility: clear the list in `thread_module_clear` so that `remove_from_shutdown_handles` is a no-op if the `thread_module_state` is cleared.

Reply:
This might be a good reason to stick with `HEAD_LOCK(_PyRuntime);` instead of a mutex in `thread_module_state` for now.

Reply:
Yeah, this is the reason that I'm using `HEAD_LOCK(&_PyRuntime)` instead of a dedicated `PyMutex` but I missed the linked list case. Nice catch!